python/lbann/contrib/lc/launcher.py

```diff
@@ -103,45 +103,57 @@ def make_batch_script(script_file=None,
 
-    # Deal with Pascal's hardware topology
-    # Note: Both GPUs on a Pascal node are on the same socket, so we
-    # only use cores on that socket.
-    if system == 'pascal' and procs_per_node == 2:
+    # Optimized thread affinity for Pascal
+    # Note: Both GPUs are on socket 0, so we only use cores on that
+    # socket.
+    if system == 'pascal':
+        cores_per_socket = cores_per_node(system) // 2
+        cores_per_proc = cores_per_socket // procs_per_node
+        environment['AL_PROGRESS_RANKS_PER_NUMA_NODE'] = procs_per_node
+        environment['OMP_NUM_THREADS'] = cores_per_proc - 1
         if scheduler == 'slurm':
-            launcher_args.append('--cpu_bind=mask_cpu:0x000001ff,0x0003fe00')
-        environment['OMP_NUM_THREADS'] = 8
-        environment['AL_PROGRESS_RANKS_PER_NUMA_NODE'] = 2
+            masks = [2**cores_per_proc - 1]
+            while len(masks) < procs_per_node:
+                masks.append(masks[-1] << cores_per_proc)
+            mask_str = ','.join([hex(mask) for mask in masks])
+            launcher_args.append('--cpu_bind=mask_cpu:{}'.format(mask_str))
 
     # Hacked bugfix for MPI_Init in MVAPICH2-2.3
     # Note: MPI_Init hangs when started with more than 35
     # processes. This bug is not present in MVAPICH2-2.2 but is
     # present in MVAPICH2-2.3rc2.
     environment['MV2_USE_RDMA_CM'] = 0
 
-    # Magic default arguments to jsrun/etc.
-    # Note: Pack processes using ten cores for each, with 40 cores total, and
-    # all four GPUs visible to each process.
+    # Optimizations for Sierra-like systems
     if system in ('sierra', 'lassen'):
+
+        # Set thread affinity
+        # Note: Aluminum's default thread affinity is incorrect since
+        # hwloc treats GPUs as NUMA domains.
+        # Note: There are actually 22 cores/socket, but it seems that
+        # powers of 2 are better for performance.
+        cores_per_socket = 16
+        procs_per_socket = (procs_per_node + 1) // 2
+        cores_per_proc = cores_per_socket // procs_per_socket
+        environment['AL_PROGRESS_RANKS_PER_NUMA_NODE'] = procs_per_socket
+        environment['OMP_NUM_THREADS'] = cores_per_proc
         if scheduler == 'lsf':
-            launcher_args.extend([
-                '--launch_distribution packed',
-                '--bind "packed:10"',
-                '--rs_per_host 1',
-                '--cpu_per_rs 40',
-                '--gpu_per_rs 4'
-            ])
-        environment['OMP_NUM_THREADS'] = 4
-        # Deal with topology mis-identification on Sierra/Lassen.
-        environment['AL_PROGRESS_RANKS_PER_NUMA_NODE'] = 2
+            launcher_args.append('--bind packed:{}'.format(cores_per_proc))
+
+        # Hack to enable process forking
+        # Note: OpenMPI and InfiniBand sometimes fail if an MPI
+        # process is forked. See
+        # https://www.open-mpi.org/faq/?category=openfabrics#ofa-fork.
+        environment['IBV_FORK_SAFE'] = 1
```

Comment:
Looks like this has some performance penalty (per Adam's email). Can this be optional?

Reply:
The [documentation](https://linux.die.net/man/3/ibv_fork_init) says:

> Calling ibv_fork_init() will reduce performance due to an extra system call for every memory registration, and the additional memory allocated to track memory regions. The precise performance impact depends on the workload and usually will not be significant.

This function is a convenience wrapper around the `lbann.run` function, so perhaps not using `IBV_FORK_SAFE` should be considered as expert usage?

Reply:
As long as it's clearly communicated to the user about the potential performance loss and there is a way to turn it off, I think that's fine.